DS-4317 bundles in rest

KevinVdV opened the pull request:
This is the REST API development for: https://github.com/DSpace/Rest7Contract/pull/72

The PR removes the bundle as a property of the bitstream, and rather introduces it as a HAL link
The bundles endpoint is introduced
The items endpoint will offer bundles rather than direct access to bitstreams

More details can be found in the JIRA: https://jira.duraspace.org/browse/DS-4317

jpelgrims-atmire committed:
Add initial class structure for bundles endpoint

jpelgrims-atmire committed:
Replace bitstreams in ItemRest with bundles

The list of bitstreams in ItemRest is now replaced with a list of
bundles. The bitstreams can now be accessed through the bundles.

jpelgrims-atmire committed:
Finish GET endpoints

jpelgrims-atmire committed:
Add integration tests for bundle GET endpoint

jpelgrims-atmire committed:
Add utility class for URL mapping Regexes

Added a utility class so that the URL mapping regexes (mostly used
in RestResourceController) can be reused in other controllers.

jpelgrims-atmire committed:
Fix tests for recent item changes

The item matcher didn't match anymore because it was still looking
for the bitstream link, which isn't there anymore because the item
only holds bundles now (see commit 48ea95bc). The tests were changed
to look for a bundle link.

Other than that a few tests were also specifically looking for a
bitstream link to see if a bitstream was correctly created. These
links were changed to bundle links, since a new bitstream should
be added to a new bundle by default (if there are no bundles yet)
during creation.

jpelgrims-atmire committed:
Add endpoint to add bundle to item

YanaDePauw committed:
64495: Add unit tests for bundle + fix issues

YanaDePauw committed:
64498: Add test to retrieve bistreams from bundle

YanaDePauw committed:
Merge branch 'master' into DS-4317_bundles-in-REST

YanaDePauw committed:
64639: Support creation of new bitstreams in bundles

YanaDePauw committed:
64499: Add bundle link to bitstream

YanaDePauw committed:
64638: Retrieving and updating bundle of a bitstream

YanaDePauw committed:
65105: Implement feedback

Atmire-Kristof committed:
64640: Move PATCH request for bitstreams in bundles

Atmire-Kristof committed:
64640: BundleRestRepository IT, JavaDocs and CheckStyle fixes

YanaDePauw committed:
65188: Implement feedback

YanaDePauw committed:
65188: Add typedocs

Atmire-Kristof committed:
64640: Message fix + Missing JavaDocs

abollini reviewed (commented):
I like to share some quick finding as I'm not sure to be able to provide a more comprehensive review in the next weeks

KevinVdV committed:
[DS-4317] Bundles in rest: Processing feedback

KevinVdV commented:
Processed the initial comments, let me know if you have further feedback.

cwilper reviewed (commented):
I have downloaded and tested via HAL browser successfully. It looks good, but I assume some angular changes will need to be made to cope with this new model -- is there an angular branch in the works? Would be good to coordinate the eventual merge of this with that.

(Edited Oct 31: omitted the part about projections; not using this branch as a base for that work)

KevinVdV committed:
Merge branch 'master' into DS-4317_bundles-in-REST

KevinVdV committed:
[DS-4317] Bundles in rest: Merging lasted master style fixes

benbosman reviewed (approved):
I've reviewed the code, tested some of the functionally locally, and verified the ITs to ensure all use cases are covered.

I included an inline comment, but overall it looks good to me.

tdonohue reviewed (commented):
A few quick comments. I have not completed a full review, but I'm concerned that we are now moving in a direction of creating Controllers for all Resources (which is the opposite direction from what we previously established).

MarieVerdonck committed:
Code for moving bitstreams to a different bundle moved to BundleService

KevinVdV committed:
[DS-4317] Bundles in rest: Renaming method that moves bitstreams in a bundle

KevinVdV commented:
@tdonohue I altered the method name as you requested.

On the "BitstreamController" there are 2 methods in this controller:
* getBundle() This one cannot be moved to the repository as the Bundle object isn't in the BitstreamRest class. Adding the bundle there isn't possible because this bundle object would in turn contain the bitstreams, which then would contain the bundle, ...
* move() This one cannot be moved as the RestResourceController does not have support for a PUT with a text/uri-list and a UUID.

KevinVdV committed:
[DS-4317] Bundles in rest: Fixing style issue & renaming the "BitstreamController" to the "BitstreamBundleController" as the RequestMapping maps to /api/content/bitstreams/uuid/bundle and not /api/content/bitstream/uuid

KevinVdV committed:
[DS-4317] Bundles in rest: Moving BitstreamBundleController to the existing BitstreamRestController

KevinVdV commented:
@tdonohue Just realized that we can remove the BitstreamController & move the code to the existing BitstreamRestController. Made this small change.

tdonohue commented:
@KevinVdV : What about the new `BundleController` and `ItemController` you are creating in this PR?  Are they necessary, or could they be replaced by the corresponding `*RestRepository` and `*Rest` classes, or somehow limited in their scope?  Any PRs that create new Controllers are a "red flag" to me, as it's not the normal processes we use.  However, we *do* make exceptions for specific subresources that cannot be implemented via `*RestRepository` and `*Rest`.  

For example, take a look at how the `/items` endpoint currently works.  The primary methods on that endpoint all use the `ItemRestRepository` and `ItemRest` classes (i.e. there is no `ItemController` in our codebase yet)

However, for the `/api/core/items/{uuid}/bitstreams` subresource we have a custom controller for that POST: https://github.com/DSpace/DSpace/blob/master/dspace-server-webapp/src/main/java/org/dspace/app/rest/ItemUploadController.java

Additionally, for the `/api/core/items/{uuid}/owningCollection` subresource, we have a separate custom controller: https://github.com/DSpace/DSpace/blob/master/dspace-server-webapp/src/main/java/org/dspace/app/rest/ItemOwningCollectionUpdateRestController.java

KevinVdV commented:
**BundleController**

This controller is uploading a file to "/api/core/bundles/{uuid}/bitstreams", the RestResourceController only has support to upload to the following:
* /api/core/bundles/{uuid}
* /api/core/bundles/{integer}
* /api/core/bundles

So we needed a separate controller for this for that reason.

**ItemController**

This controller is adding a bundle using the following endpoint: "/api/core/items/{uuid}/bundles"

The RestResourceController doesn't offer us this option, hence the separate controller.

We can easily rename both controllers to:
* BundleUploadBitstreamController
* ItemAddBundleController

But I think we need to do some refactoring, as the number of controllers in the "org.dspace.app.rest" package will keep increasing. I would propose to rename them & move them to the "org.dspace.app.rest.controller" package. When this is in master we can then move the other custom controllers in this package as well.

Thinking a bit further on this, we could go further & create an "org.dspace.app.rest.controller.content" package, move all controllers related to the api "content" package in this package. Then create an "org.dspace.app.rest.controller.discovery" package & move the custom Discovery controllers in this package, etc...

What do you think ?

tdonohue commented:
@KevinVdV : Thanks for the explanation here. I agree both Controllers should be **renamed**... their current names imply they are implementing all the endpoints under `/items` and `/bitstreams`, which they obviously are not doing.  Once they are renamed (and JavaDocs added to each to describe them, as currently they don't have JavaDocs), then I think this is fine to move forward.  I think your suggestions for new names are fine.

I also agree with your concern about the large number of controllers under the `org.dspace.app.rest` package. I like the suggestion to simply move them to `org.dspace.app.rest.controller` for now.  I don't know that we need to subpath that further though, as we haven't done so in other areas (like `org.dspace.app.rest.repository` or `org.dspace.app.rest.model`.  However, I'd recommend we move these controllers in a separate PR (all at once).  That's a PR I could quickly review & merge immediately as it'd just involve moving files around (and correcting import statements).

Thanks @KevinVdV !

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/BundleController.java:
```diff
@@ -0,0 +1,99 @@
+
+@RestController
+@RequestMapping("/api/core/bundles/{uuid}")
+public class BundleController {
```
This Controller is missing JavaDocs and should be renamed (@KevinVdV suggested `BundleUploadBitstreamController`) as it does NOT implement the entire `/bundles` endpoint (as the name suggests).

I'd also recommend correcting the `@RequestMapping` to be more specific, along the lines of `"/api/core/bundles/{uuid}/bitstreams"`

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/ItemController.java:
```diff
@@ -0,0 +1,92 @@
+
+@RestController
+@RequestMapping("/api/core/items" + REGEX_REQUESTMAPPING_IDENTIFIER_AS_UUID + "/bundles")
+public class ItemController {
```
This Controller is missing JavaDocs and should be renamed (@KevinVdV suggested ItemAddBundleController) as it does NOT implement the entire `/items` endpoint (as the name suggests).

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/BitstreamRestController.java:
```diff
@@ -241,4 +264,87 @@ public BitstreamResource updateBitstreamFormat(@PathVariable UUID uuid,
+     */
+    @PreAuthorize("hasPermission(#uuid, 'BITSTREAM', 'READ')")
+    @RequestMapping(method = {RequestMethod.GET, RequestMethod.HEAD}, value = "/bundle")
+    public ResponseEntity<ResourceSupport> getBundle(@PathVariable UUID uuid, HttpServletResponse response,
```
A part of me wonders if these `/bundle` methods should move to a different controller named `BitstreamBundleController`, as the Javadocs of the `BitstreamRestController` state the the controller is "to provide access to the bitstream binary content" (and bundles don't fall under that purpose).  At a minimum, we need to update the JavaDocs here to change the scope of this Controller.

In general, I like the idea of having smaller, more descriptive controllers overall, rather than a generically named controller....as it makes it easier to clarify the scope of what is implemented in the Controller vs what is provided by the `RestRepository` classes.

However, that said, I'm OK with this change for now (if JavaDocs are updated).  We can always refactor later if deemed necessary.

tdonohue reviewed (commented):
@KevinVdV : Per our discussion above, I've added in inline comments on the necessary changes I see. Once these are made, I think this is good to go.

MarieVerdonck committed:
Bundles in REST: Feedback implemented:
 - Renaming: BundleController > BundleUploadBitstreamController, ItemController > ItemAddBundleController
 - Moved: BitstreamRestController#getBundle() & #move() to new BitstreamBundleController class
 - Javadocs added to all three
 - Correction @RequestMapping of BundleUploadBitstreamController

MarieVerdonck committed:
Merge branch 'master' into DS-4317_bundles-in-REST

# Conflicts:
#	dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/BitstreamRestRepository.java

KevinVdV commented:
@tdonohue We have made the changes you requested. Can you review again ?

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/BitstreamBundleController.java:
```diff
@@ -0,0 +1,159 @@
+ */
+@RestController
+@RequestMapping("/api/" + BitstreamRest.CATEGORY + "/" + BitstreamRest.PLURAL_NAME
+        + REGEX_REQUESTMAPPING_IDENTIFIER_AS_UUID)
```
RequestMapping should be `@RequestMapping("/api/" + BitstreamRest.CATEGORY + "/" + BitstreamRest.PLURAL_NAME + REGEX_REQUESTMAPPING_IDENTIFIER_AS_UUID + "/bundle")`  as ALL methods here are for the `/bundle` subpath.  

This also ensures we don't accidentally map *two controllers to the same path*....keep in mind the generic `RestResourceController` will already be mapping `/api/core/bitstreams` related paths.  See this comment on another PR (which has the same mistake) for more detail: https://github.com/DSpace/DSpace/pull/2529#discussion_r346893787

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/BundleUploadBitstreamController.java:
```diff
@@ -0,0 +1,116 @@
+ * </pre>
+ */
+@RestController
+@RequestMapping("/api/core/bundles/" + REGEX_REQUESTMAPPING_IDENTIFIER_AS_UUID)
```
Same as above, this RequestMapping should be *more specific*.  It should be: `@RequestMapping("/api/core/bundles/" + REGEX_REQUESTMAPPING_IDENTIFIER_AS_UUID + "/bitstreams")` as it is *specifically* for `/api/core/bundles/{uuid}/bitstreams` requests.

As noted in the above comment, this also ensures we are not creating *two controllers* with the same RequestMapping.  Doing so is NOT recommended as it can result in confusing (and difficult to track down bugs) in the future.  Again, see comment above for more info too.

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/ItemAddBundleController.java:
```diff
@@ -0,0 +1,114 @@
+ */
+@RestController
+@RequestMapping("/api/core/items" + REGEX_REQUESTMAPPING_IDENTIFIER_AS_UUID + "/bundles")
+public class ItemAddBundleController {
```
Thank you! This is great example of a correct controller with a more specific requestmapping

tdonohue reviewed (commented):
@KevinVdV : This is _very close_, but we still seem to be having issues with creating controllers that duplicate the `RequestMapping` of existing Controllers.  We need to find a way to avoid that, as I'll be checking closely for that in every code review I do.  This is NOT recommended in Spring development (in general) as it can result in odd bugs later on if a developer ever adds duplicative methods to those duplicative mapped controllers.  

I'd strongly prefer we never have Controllers that duplicate the same `RequestMapping` (that said, I think there are a few already in the codebase that need cleaning up...let's just avoid it in future PRs, please.)

MarieVerdonck committed:
DS-4317 - bundles in REST: Redefining RequestMapping of Controllers to be unique and to use constants defined in Rest classes

MarieVerdonck committed:
DS-4317 - bundles in REST: Redefined RequestMapping to use constants defined in Rest classes

tdonohue reviewed (approved):
👍 Looks good now, thanks for the updates!
